pkg/virt-handler/vm.go

```diff
@@ -1380,6 +1386,56 @@ func (d *VirtualMachineController) handleMigrationProxy(vmi *v1.VirtualMachineIn
 	return nil
 }
 
+// This function does just enough to get a domain object that would allow
+// phase1 network configuration to succeed. This object is not complete and may
+// miss some important data useful in a different context. Be cautious before
+// using the result of this function in a different context. Consider using
+// domainInformer instead as available.
+func getDomain(vmi *v1.VirtualMachineInstance) (*api.Domain, error) {
+	isBlockPVCMap := make(map[string]bool)
+	isBlockDVMap := make(map[string]bool)
+	diskInfo := make(map[string]*containerdisk.DiskInfo)
+	for _, volume := range vmi.Spec.Volumes {
+		if volume.VolumeSource.PersistentVolumeClaim != nil {
+			isBlockPVCMap[volume.Name] = true
+		} else if volume.VolumeSource.ContainerDisk != nil {
+			diskInfo[volume.Name] = &containerdisk.DiskInfo{}
+		} else if volume.VolumeSource.DataVolume != nil {
+			isBlockDVMap[volume.Name] = true
+		}
+	}
+
+	c := &api.ConverterContext{
+		VirtualMachine: vmi,
+		DiskType:       diskInfo,
+		IsBlockPVC:     isBlockPVCMap,
+		IsBlockDV:      isBlockDVMap,
+		// virt-handler may not have kvm initialized
+		UseEmulation: true,
+		// skip non-essential initialization that requires full
+		// virt-launcher execution context (envvars etc.)
+		BestEffort:   true,
+	}
+	domain := &api.Domain{}
+	if err := api.Convert_v1_VirtualMachine_To_api_Domain(vmi, domain, c); err != nil {
+		return nil, err
+	}
+	return domain, nil
+}
+
+func setupNetworkPhase1(vmi *v1.VirtualMachineInstance) error {
+	// domainInformer doesn't have the domain cached just yet, so we have
+	// to reconstruct the object here
+	domain, err := getDomain(vmi)
```

Comment:
We should make as little as possible inside the netns. Could you move this helper function outside?

Reply:
@phoracek what's the reason to do as little as possible inside the netns? Are there any races in the latest golang implementations I am not aware of? Otherwise, I find isolating domain extraction business in this helper function keeps the caller somewhat cleaner from the dirty domain business details.